src/VisualStudio/Core/Def/ProjectSystem/MetadataReferences/VisualStudioMetadataReferenceManager.cs

```diff
@@ -120,44 +131,45 @@ internal Metadata GetMetadata(string fullPath, DateTime snapshotTimestamp)
             var key = new FileKey(fullPath, snapshotTimestamp);
             // check existing metadata
             if (_metadataCache.TryGetMetadata(key, out var metadata))
-            {
                 return metadata;
-            }
 
-            if (VsSmartScopeCandidate(key.FullPath))
-            {
-                var newMetadata = CreateAssemblyMetadataFromMetadataImporter(key);
-                var metadataValueSource = ValueSource.Constant(newMetadata);
-                if (!_metadataCache.GetOrAddMetadata(key, metadataValueSource, out metadata))
-                {
-                    newMetadata.Dispose();
-                }
+            var (newMetadata, newMetadataValueSource) = GetMetadataWorker();
```

Comment:
the diff looks bad.  but here's the concept:

1. the code here moved to a local function that now returns both the metadata and hte value-source wrapping the metadata.
2. this was teh logic to update the cache, and dispose of the metadata if we fail, is in one place, not multiple.
3. the logic moved into the local function is about hte same, except it now checks the A/B flag to see if we should bother with recoverable metadata, or just use a ConstantValueSoruce.